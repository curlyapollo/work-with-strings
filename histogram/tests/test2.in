#include <exception>
#include <functional>
#include <optional>

template <typename Result, typename Exception = std::exception>
std::optional<Result> DoWithRetry(std::function<Result()> func,
                                  int retryCount, int sleepTime, bool throwLast) {
    auto counter = 0;
    do {
        try {
            auto s = func();
            return s;
        }
        catch (Exception e) {
            counter++;
            if (counter == retryCount + 1) {
                if (throwLast) {
                    throw std::move(e);
                }
                else
                    return std::optional<Result>();
            }
            Sleep(sleepTime);
        }
    } while (counter != retryCount + 1);
    return Result();
}
