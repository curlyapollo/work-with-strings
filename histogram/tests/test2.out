if 2
else 1
for 0
int 2
void 0
